### PR TITLE
Refactor Diagram.js to use hooks

### PR DESCRIPTION
**Pull request by husseyexplores**
#### What does this PR do?
This PR refactors Diagram.js to use hooks.

#### Where should the reviewer start?
`/src/js/components/Diagram/Diagram.js`

#### What testing has been done on this PR?
Storybook

#### How should this be manually tested?
Storybook

#### What are the relevant issues?
#3394 

#### Do the grommet docs need to be updated?
No

#### Should this PR be mentioned in the release notes?
No

#### Is this change backwards compatible or is it a breaking change?
Backwards compatible

### CODE REVIEW

File: src/js/components/Diagram/Diagram.js [hunk 1]
```diff
@@ -165,75 +141,83 @@ class Diagram extends Component {
...
+      viewBox={`0 0 ${width} ${height}`}
+      preserveAspectRatio="xMinYMin meet"
+      {...rest}
+    >
+      <g>{paths}</g>
+    </StyledDiagram>
+  );
+};
+
+Diagram.defaultProps = { connections: [] };
```
Comment (ShimiSun) [inline src/js/components/Diagram/Diagram.js]: Can we please move this assignment to line #63 on the component declaration?

Comment (husseyexplores) [reply src/js/components/Diagram/Diagram.js]: Made the change.

**Comment by ShimiSun** [comment]
Hi buddy, do you mind taking a look at the failed test, 

```
 FAIL  src/js/components/__tests__/components-test.js
  ● Components loads

    expect(received).toMatchSnapshot()

    Snapshot name: `Components loads 1`

    - Snapshot
    + Received

    @@ -1762,11 +1762,10 @@
            Diagram can be used to draw lines connecting the Boxes.",
          "intrinsicElement": "svg",
          "name": "Diagram",
          "properties": Array [
            Object {
    -         "defaultValue": Array [],
              "description": "Array of objects describing the connections.
            The 'fromTarget' and 'toTarget' may be either DOM element ids or
            React references.
            'offset' can be used to shift a bit to reduce the amount of overlap
            with other connection lines to make the lines easier to distinguish.",

      2 | 
      3 | test('Components loads', () => {
    > 4 |   expect(Components).toMatchSnapshot();
        |                      ^
      5 | });
      6 | 

      at Object.toMatchSnapshot (src/js/components/__tests__/components-test.js:4:22)

 › 1 snapshot failed.
Summary of all failing tests
 FAIL  src/js/components/__tests__/components-test.js
  ● Components loads

    expect(received).toMatchSnapshot()

    Snapshot name: `Components loads 1`

    - Snapshot
    + Received

    @@ -1762,11 +1762,10 @@
            Diagram can be used to draw lines connecting the Boxes.",
          "intrinsicElement": "svg",
          "name": "Diagram",
          "properties": Array [
            Object {
    -         "defaultValue": Array [],
              "description": "Array of objects describing the connections.
            The 'fromTarget' and 'toTarget' may be either DOM element ids or
            React references.
            'offset' can be used to shift a bit to reduce the amount of overlap
            with other connection lines to make the lines easier to distinguish.",

      2 | 
      3 | test('Components loads', () => {
    > 4 |   expect(Components).toMatchSnapshot();
        |                      ^
      5 | });
      6 | 

      at Object.toMatchSnapshot (src/js/components/__tests__/components-test.js:4:22)


Snapshot Summary
 › 1 snapshot failed from 1 test suite. Inspect your code changes or run `yarn test -u` to update them.

Test Suites: 1 failed, 51 passed, 52 total
Tests:       1 failed, 447 passed, 448 total
Snapshots:   1 failed, 557 passed, 558 total
Time:        40.779s
```
https://circleci.com/gh/grommet/grommet/8682?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by husseyexplores** [comment]
Sorry, I overlooked the failing test.

It seems weird though. The test failed when I moved default props declaration from a separate block (`Diagram.defaultProps = { connections: [] };`) to the the function parameters (`Diagram = ({ connectios = [], ... })`.
You can see above that my [commit](https://github.com/grommet/grommet/pull/3458/commits/4aa073db5fbb6c6d55d0aaf83d1266012b469044) to refactor the default props results in a failing build.

I also made a couple of minor changes that I overlooked in my latest commit. i.e using `useCallback` to wrap the callback handler and using ref to access the latest state. But this does not affect any test to fail or pass.

I have no idea why the test is failing if we declare the default props in function parameters. Any ideas? Seems an issue with tests to me because declaring default props in function parameters is indeed a [valid way](https://twitter.com/dan_abramov/status/1133878326358171650).

**Comment by husseyexplores** [comment]
Another thing is that we can extract some of the functionality into custom hooks?
I'm especially talking about extracting add/remove event handler into possibly a `useEventListener` hook.
Example: https://codesandbox.io/s/z64on3ypm

And also adding `eslint-plugin-react-hooks` into the project will definitely be worth it since we are converting the codebase to use hooks.

**Comment by ShimiSun** [comment]
>Another thing is that we can extract some of the functionality into custom hooks?
I'm especially talking about extracting add/remove event handler into possibly a useEventListener hook.

Looks good! we can start with a PR and see how it goes. the one thing to keep in mind is that we want to keep the code readable for the open-source community.

>And also adding eslint-plugin-react-hooks into the project will definitely be worth it since we are converting the codebase to use hooks.

Great idea! go for it!

File: src/js/components/Diagram/Diagram.js [hunk 2]
```diff
@@ -165,75 +162,74 @@ class Diagram extends Component {
...
```
Comment (ShimiSun) [inline src/js/components/Diagram/Diagram.js]:
```suggestion
Diagram.defaultProps = {};
```
moving the initialization to the declaration point

File: src/js/components/Diagram/Diagram.js [hunk 3]
```diff
@@ -60,59 +60,56 @@ const findTarget = target => {
...
-    window.addEventListener('resize', this.onResize);
-    this.onResize();
-  }
-
-  componentDidUpdate() {
-    const { connectionPoints } = this.state;
-    if (!connectionPoints) {
-      this.placeConnections();
-    }
+const Diagram = ({ connections, theme, ...rest }) => {
```
Comment (ShimiSun) [inline src/js/components/Diagram/Diagram.js]:
```suggestion
const Diagram = ({ connections = [], theme, ...rest }) => {
```
moving up from default props

Comment (husseyexplores) [reply src/js/components/Diagram/Diagram.js]: I just did it again anyway. It fails a test in [src/js/components/__tests__](https://github.com/grommet/grommet/blob/master/src/js/components/__tests__/components-test.js)

I have no idea why.. I mentioned this above also.

Comment (ShimiSun) [reply src/js/components/Diagram/Diagram.js]: yea, you are right, sorry for the hustle. I'll need to take a look at it myself, can you roll back your PR to have the array with the default props?

Comment (husseyexplores) [reply src/js/components/Diagram/Diagram.js]: Pushed a new commit, I don't know how to rollback a commit. I'm a very newbie!

Comment (ShimiSun) [reply src/js/components/Diagram/Diagram.js]: no worries, you are doing great my friend, at the end, it is all squashed anyway :)

File: src/js/components/Diagram/Diagram.js [hunk 4]
```diff
@@ -60,59 +60,56 @@ const findTarget = target => {
...
+  const [width, setWidth] = useState(0);
+  const [height, setHeight] = useState(0);
```
[the remark this answers is not in the document]
Comment (husseyexplores) [reply src/js/components/Diagram/Diagram.js]: Sure, that makes sense!

File: src/js/components/Diagram/Diagram.js [hunk 5]
```diff
@@ -60,59 +60,56 @@ const findTarget = target => {
...
+  // https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops
+  if (connections !== prevConnections) {
+    setPrevConnections(connections);
+    setConnectionPoints(undefined);
   }
 
-  componentWillUnmount() {
-    window.removeEventListener('resize', this.onResize);
-  }
+  const onResize = useCallback(() => {
```
[the remark this answers is not in the document]
Comment (husseyexplores) [reply src/js/components/Diagram/Diagram.js]: `onResize` depends on `width` and `height` variables from state. So whenever these variables will change, our effect would re-run and add/remove the handler again.. and again.
If it's not much of a problem, I can move the handler into the effect.

[the remark this answers is not in the document]
Comment (ShimiSun) [reply src/js/components/Diagram/Diagram.js]: I'll extract your suggestion to be an enhancement, thank you for sharing your inputs!

**Comment by ShimiSun** [comment]
Thank you for the contribution and patient buddy! 🎉

**Pull request merged**